### ISSUE TITLE
OCSP stapling with HAProxy as a server

### ISSUE THREAD

**Issue by einval22** [wolfSSL/wolfssl#7588]
### Version

v5.7.0-stable

### Description

Hello guys ! 

We have some user [requests](https://github.com/haproxy/haproxy/issues/2535) to make work OCSP stapling extension with haproxy process as a server, i.e. haproxy  loads, parses by itself OCSP response for the given server certificate and provides the  status to clients. 

haproxy is linked with libwolfssl.so.42, compiled from v5.7.0-stable.

We have some problems to make OCSP extension work, in spite of the support, which was added in our source code via this patch 3cbf09ed6411 "MEDIUM: ssl: add minimal WolfSSL support with OpenSSL compatibility mode" in haproxy ".

We have the following workflow for processing OCSP response files in DER format:

1. haproxy starts and loads from the filesystem some server certificate and its corresponding OCSP response;
2. it calculates CertID and it sets OCSP stapling callback [here](https://github.com/haproxy/haproxy/blob/master/src/ssl_sock.c#L1214);
3. it finally parses and checks OCSP response file by itself [here](https://github.com/haproxy/haproxy/blob/master/src/ssl_ocsp.c#L213).

Problems have started already, when we try to check CertID length via wolfSSL_i2d_OCSP_CERTID helper [here](https://github.com/haproxy/haproxy/blob/master/src/ssl_sock.c#L1172).  From the  [wolfSSL_i2d_OCSP_CERTID](https://github.com/wolfSSL/wolfssl/blob/master/src/ocsp.c#L1224) code we can see, that it uses `rawCertId` and `rawCertIdSize` fields from [OcspEntry](https://github.com/wolfSSL/wolfssl/blob/master/wolfssl/wolfcrypt/asn.h#L2495) structure, which are not filled by [wolfSSL_OCSP_cert_to_id](https://github.com/wolfSSL/wolfssl/blob/master/src/ocsp.c#L724).  Could you, please, provide some explanations on this ?

When we did some workaround and  started to simply check in our code the total length of 
` OcspEntry->issuerHash + OcspEntry->issuerKeyHash + OcspEntry->status->serial < OCSP_MAX_CERTID_ASN1_LENGTH (128 bytes)`, 

we could advance further and could finally call `wolfSSL_CTX_set_tlsext_status_cb`, which triggers  the path:
` wolfSSL_CTX_EnableOCSPStapling --> wolfSSL_CertManagerEnableOCSPStapling --> cm->ocspStaplingEnabled = 1`
and sets our `ssl_sock_ocsp_stapling_cbk` callback function.

Then we could also successfully parse DER OCSP response, [wolfSSL_d2i_OCSP_RESPONSE](https://github.com/wolfSSL/wolfssl/blob/master/src/ocsp.c#L979)  works as expected.  So,  finally we've successfully loaded and checked OCSP response data, but still no OCSP extension provided  in SERVER HELLO, when a client sends CLIENT HELLO with Certificate Status Request (TLSv1.2).

From WolfSSL debug logs we see that SendCertificateStatus returns 0 in the case:

```
wolfSSL Entering wolfSSL_new
wolfSSL Entering ReinitSSL
RNG_HEALTH_TEST_CHECK_SIZE = 128
sizeof(seedB_data)         = 128
opened /dev/urandom.
rnd read...
wolfSSL Entering SetSSL_CTX
wolfSSL Entering wolfSSL_set_options
	SSL_OP_ALL
	WOLFSSL_OP_NO_SSLv2 : wolfSSL does not support SSLv2
	SSL_OP_NO_TLSv1_1
	SSL_OP_NO_TLSv1
	SSL_OP_NO_SSLv3
	WOLFSSL_OP_CIPHER_SERVER_PREFERENCE
SSL_OP_NO_COMPRESSION: compression not compiled in
wolfSSL Entering wolfSSL_NewSession
InitSSL done. return 0 (success)
wolfSSL_new InitSSL success
wolfSSL Leaving wolfSSL_new InitSSL =, return 0
wolfSSL Entering wolfSSL_BIO_new
wolfSSL Stub wolfSSL_BIO_set_init not implemented
...
wolfSSL Entering wolfSSL_accept
wolfSSL Entering ReinitSSL
wolfSSL Entering RetrySendAlert
wolfSSL Entering RetrySendAlert
wolfSSL Entering BioReceive
wolfSSL Entering wolfSSL_BIO_read
wolfSSL Entering wolfSSL_BIO_get_data
wolfSSL Entering wolfSSL_BIO_clear_retry_flags
Client attempting to connect with different version
growing input buffer
wolfSSL Entering BioReceive
wolfSSL Entering wolfSSL_BIO_read
wolfSSL Entering wolfSSL_BIO_get_data
wolfSSL Entering wolfSSL_BIO_clear_retry_flags
received record layer msg
got HANDSHAKE
wolfSSL Entering wolfSSL_get_options
wolfSSL Entering DoTls13HandShakeMsg
wolfSSL Entering EarlySanityCheckMsgReceived
wolfSSL Leaving EarlySanityCheckMsgReceived, return 0
wolfSSL Entering DoTls13HandShakeMsgType
processing client hello
wolfSSL Entering DoTls13ClientHello
wolfSSL Entering DoClientHello
    downgrading to TLSv1.2
Matched No Compression
Adding signature algorithms extension
Point Formats extension received
Supported Groups extension received
Session Ticket extension received
Certificate Status Request extension received
Encrypt-Then-Mac extension received
Extended Master Secret extension received
Signature Algorithms extension received
TLSX SigAlgo list exceeds max, truncating
wolfSSL Entering ALPN_Select
Calling user cert setup callback
wolfSSL Entering wolfSSL_get_client_suites_sigalgs
wolfSSL Entering wolfSSL_version
Unsupported cipher suite, CipherRequires
....
wolfSSL Entering wolfSSL_CTX_get_verify_mode
....
accept state ACCEPT_CLIENT_HELLO_DONE
accept state ACCEPT_FIRST_REPLY_DONE
wolfSSL Entering SendServerHello
growing output buffer
wolfSSL Entering wolfSSL_get_options
Session Ticket extension to write
Point Formats extension to write
Secure Renegotiation extension to write
EMS extension to write
wolfSSL Leaving SendServerHello, return 0
accept state SERVER_HELLO_SENT
wolfSSL Entering SendCertificate
growing output buffer
wolfSSL Leaving SendCertificate, return 0
accept state CERT_SENT
wolfSSL Entering SendCertificateStatus
wolfSSL Leaving SendCertificateStatus, return 0
```

According to the code from [SendCertificateStatus](https://github.com/wolfSSL/wolfssl/blob/master/src/internal.c#L23884), it seems, that we should enter in `case WOLFSSL_CSR2_OCSP:`.  Unfortunately, `CreateOcspResponse` does not contain any traces to simply check where it fails.

From `CreateOcspResponse` code we can see that  probably it does not suppose to exit on:

```
 /* unable to fetch status. skip. */
    if (SSL_CM(ssl) == NULL || SSL_CM(ssl)->ocspStaplingEnabled == 0)
        return 0;
```
check, as SSL_CM(ssl)->ocspStaplingEnabled was set to one.

We have an intention to make work OCSP and we could probably investigate further, but there are  some worries about the fact  that `wolfSSL_CertManagerEnableOCSPStapling` along with `ocspStaplingEnabled = 1` also sets the built-in OCSP lookup [here](https://github.com/wolfSSL/wolfssl/blob/master/src/ssl_certman.c#L2140) ([EmbedOcspLookup](https://github.com/wolfSSL/wolfssl/blob/master/src/wolfio.c#L1892)).  In this lookup you build and send OCSP request to the detected OCSP URI.  This is quite same logic, that we do on our side  in [ssl_ocsp_update_responses](https://github.com/haproxy/haproxy/blob/master/src/ssl_ocsp.c#L887). 

So could you please, confirm, that, if we reset `cm->ocspIOCb` and `cm->ocspRespFreeCb` to NULL, as we update OCSP responses by ourselves, we do not have any other drawbacks and we are on the right way to make things work.  And could you also provide to us some hints,  where potentially `SendCertificateStatus` or `CreateOcspResponse` may fail.

Many thanks in advance,
Kind regards.

**Comment by anhu** [comment]
Hi @vkssv , 
Thank you so much for all the details and work that you have put into this issue you are having.  Sorry our response time is a little slow as this is an American holiday today so most of our staff is out. 

I see your configuration command is ` ./configure --enable-haproxy --enable-debug --enable-quic --prefix=/opt/wolfssl-5.7.0` and you are using wolfSSL 5.7.0.  Can you give exact steps to reproduce?  

For example, how are you building and executing haproxy?  Do I need any certificates or keys to reproduce?  I see the peer is openSSL s_client.  Does it need a CA certificate? 

This might be much simpler if we can reproduce this using our `examples/server/server` and `examples/client/client` and that is why I am asking about all these artifacts.  

Its great to see haproxy/wolfssl interest on constrained embedded devices.  Lets see if we can get this working!

**Comment by anhu** [comment]
Hi @vkssv , 

Does https://github.com/wolfSSL/wolfssl/pull/7593 fix this issue? 

Warm regards, Anthony

**Comment by dgarske** [comment]
@rizlik please provide an update to @vkssv . Thanks

**Comment by hpn0t0ad** [comment]
Hi! It's unfortunate that there doesn't seem to be much interest in making wolfssl work correctly with haproxy. I believe that many haproxy users would happily ditch quictls/openssl for wolfssl (myself included) but can't, for lack of ocsp stapling support. Is there any way users could help debug this issue?

**Comment by dgarske** [comment]
Hi @hpn0t0ad , thank you for the reminder about this issue. The engineer I assign is on vacation. Let me find another engineer to work on this. Thanks, David Garske, wolfSSL

**Comment by ColtonWilley** [comment]
Hello @vkssv 

I was looking at this issue to reproduce, however it seems the certificates you provided (testocsp.irq6.net.pem) has expired. I was hoping you could provide me with details on how you created it and the ocsp file so I can test on my end. 

Thanks,
Colton Willey, wolfSSL.

**Comment by wlallemand** [comment]
Just updating the ticket here for people interested by the feature, there is a PR https://github.com/wolfSSL/wolfssl/pull/8036 which should fix part of the issues, we could probably have the feature soon in haproxy once this is merged.

**Comment by hpn0t0ad** [comment]
Hi guys! Now that 268326d875c0020bfb626f23c3f323ecfcffe104 has been merged does that mean we finally have working support for OCSP stapling with haproxy? If so, do we have to do anything special to enable it? --enable-haproxy --enable-ocsp --enable-ocspstapling --enable-ocspstapling2 ?

**Comment by dgarske** [comment]
> Hi guys! Now that [268326d](https://github.com/wolfSSL/wolfssl/commit/268326d875c0020bfb626f23c3f323ecfcffe104) has been merged does that mean we finally have working support for OCSP stapling with haproxy? If so, do we have to do anything special to enable it? --enable-haproxy --enable-ocsp --enable-ocspstapling --enable-ocspstapling2 ?

Hi @hpn0t0ad , As far as I know that's all you would need. I'll defer to @rizlik for tips on this. Are you seeing any build errors or issues? Thanks, David Garske, wolfSSL

**Comment by hpn0t0ad** [comment]
> Hi [@hpn0t0ad](https://github.com/hpn0t0ad) , As far as I know that's all you would need. I'll defer to [@rizlik](https://github.com/rizlik) for tips on this. Are you seeing any build errors or issues? Thanks, David Garske, wolfSSL

No build errors with current master, but I'm still seeing the same error as always logged by haproxy:
```
[ALERT]    (10) : config : parsing [/usr/local/etc/haproxy/haproxy.cfg:41] : 'bind quic4@:443' in section 'listen' : 'crt-list' : error processing line 6 in file '/usr/local/etc/haproxy/crt.list' :  '/usr/local/etc/haproxy/certs/go-foo-com-chain-key.pem' has an OCSP auto-update set to 'on' but an error occurred (maybe theOCSP URI or the issuer could not be found)'.
```
I have no idea how to debug this further but suspect that there is still some work on the haproxy side to be done to fully support this feature with wolfssl.

**Comment by wlallemand** [comment]
@hpn0t0ad hello, I didn't had time to test the PR yet, there might be some things to change in HAProxy. I'll update the ticket once we checked this.

**Comment by wlallemand** [comment]
@hpn0t0ad I checked the current WolfSSL master, and it seems like there are still some things in the API that prevents the OCSP code in HAProxy to work. The OCSP callback is now working, but it still lacks other functions that are needed:


In HAProxy we need to load the CERTID which is used as a key in a tree before calling the callback. Which is done with something like this:           
                                                                                                                               
        unsigned int len;                                                                                                      
        unsigned char *p = NULL;                                                                                               
        OCSP_CERTID *cid = OCSP_cert_to_id(0, x, issuer);                                                                      
        len = i2d_OCSP_CERTID(cid, NULL);                                                                                      
        if (!len || (len > OCSP_MAX_CERTID_ASN1_LENGTH))                                                                       
            goto out;                                                                                                          
        len = i2d_OCSP_CERTID(cid, &p);                                                                                        
                                                                                                                               
But it seems like the i2d_OCSP_CERTID() function always returns 0 so we can't go further in the code.
see there: https://github.com/haproxy/haproxy/blob/master/src/ssl_sock.c#L1174

**Comment by rizlik** [comment]
@wlallemand , @hpn0t0ad 

I think I identified the issue and working on a fix. I'll let you know as soon as I put up a PR.

**Comment by rizlik** [comment]
@wlallemand , @hpn0t0ad 
Fixed were merged, HAproxy can be tested again.

**Comment by hpn0t0ad** [comment]
Not working for me with latest master and haproxy 3.1.5 unfortunately. I'm not sure how to get more verbose debug output from haproxy; all I can tell is that with openssl I get
`<133>Mar  3 13:59:16 haproxy[10]: <OCSP-UPDATE> /usr/local/etc/haproxy/certs/go-foo-com-chain-key.pem 1 "Update successful" 0 1`
while with wolfssl it fails with
`[ALERT]    (10) : config : parsing [/usr/local/etc/haproxy/haproxy.cfg:41] : 'bind quic4@:443' in section 'listen' : 'crt-list' : error processing line 6 in file '/usr/local/etc/haproxy/crt.list' :  '/usr/local/etc/haproxy/certs/go-foo-com-chain-key.pem' has an OCSP auto-update set to 'on' but an error occurred (maybe the OCSP URI or the issuer could not be found)'.`.

I've configured wolfssl with
```
--enable-haproxy \
--enable-quic \
--enable-ocsp \
--enable-ocspstapling \
--enable-ocspstapling2 \
```
and haproxy with
```
USE_QUIC=1 \
USE_OPENSSL_WOLFSSL=1 \
```

**Comment by rizlik** [comment]
Hi, we did several changes in the OCSP so that HAproxy should work now. Let us know if you have any other problems. 
Feel free to open another issue or to send a mail to support@wolfssl.com otherwise.
